### PR TITLE
[SHELL32][SHELL32_APITEST][SDK] Implement SheRemoveQuotesA/W

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Implementing missing features...
JIRA issue: [CORE-9277](https://jira.reactos.org/browse/CORE-9277)

## Proposed changes

- Implement `SheRemoveQuotesA` and `SheRemoveQuotesW` functions.
- Add `She` testcase into `shell32_apitest.exe`.
- Add `SheRemoveQuotesA` and `SheRemoveQuotesW` prototypes into `undocshell.h`

## TODO

- [x] Do small tests.

## Comparison

WinXP:
WinXP
Successful.

Win2k3:
Win2k3
Successful.

Win10:
Win10
Skipped because there are no `SheRemoveQuotesA/W` functions.

ReactOS (AFTER):
after
Successful.

### CODE REVIEW

**Review by JoachimHenze — APPROVED** [review]
Approved. Can confirm those funcx are exported even on XPSP3 already.
Implementation does look straigthforward to me.
Question: Where do you intend to use them Katayama?

**Comment by katahiromz** [comment]
> Question: Where do you intend to use them Katayama?

This function is used in `shell32!ExtractAssociatedIconExW`.

File: dll/win32/shell32/stubs.cpp [hunk 1]
```diff
@@ -451,26 +451,38 @@ SheSetCurDrive(INT iIndex)
...
 EXTERN_C LPWSTR
 WINAPI
-SheRemoveQuotesW(LPWSTR lpInput)
+SheRemoveQuotesW(LPWSTR psz)
 {
-    FIXME("SheRemoveQuotesW() stub\n");
-    return NULL;
+    PWCHAR pch;
+
+    if (*psz == L'"')
```
Comment (JoachimHenze) [inline dll/win32/shell32/stubs.cpp:460]:
```suggestion
    if (psz && *psz == L'"')
```
Wouldn't your initial implemenation crash when being called with a nullptr? I would recommend to add a test with NULL being passed also.

File: dll/win32/shell32/stubs.cpp [hunk 2]
```diff
@@ -451,26 +451,38 @@ SheSetCurDrive(INT iIndex)
...
 EXTERN_C LPSTR
 WINAPI
-SheRemoveQuotesA(LPSTR lpInput)
+SheRemoveQuotesA(LPSTR psz)
 {
-    FIXME("SheRemoveQuotesA() stub\n");
-    return NULL;
+    PCHAR pch;
+
+    if (*psz == '"')
```
Comment (JoachimHenze) [inline dll/win32/shell32/stubs.cpp:477]:
```suggestion
    if (psz && *psz == '"')
```

Wouldn't your initial implemenation crash when being called with a nullptr? I would recommend to add a test with NULL being passed also.

**Review by JoachimHenze — CHANGES_REQUESTED** [review]

File: dll/win32/shell32/stubs.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/stubs.cpp:460]: The original function doesn't check it.

File: dll/win32/shell32/stubs.cpp (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shell32/stubs.cpp:477]: The original function doesn't check it.

File: dll/win32/shell32/stubs.cpp (hunk 1 above)
Comment (JoachimHenze) [reply dll/win32/shell32/stubs.cpp:460]: So you say if you pass a nullptr to that func in Windows it will crash?
Does that still apply to Windows 10?

Comment (katahiromz) [reply dll/win32/shell32/stubs.cpp:460]: Added `NULL` tests.
Confirmed in Win2k3 and WinXP.
Win10 skips the tests.

File: dll/win32/shell32/stubs.cpp (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shell32/stubs.cpp:477]: Added `NULL` tests.
Confirmed in Win2k3 and WinXP.
Win10 skips the tests.

File: modules/rostests/apitests/shell32/She.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    CHAR sz2[] = "\"Test\"123";
+
+    BOOL bGotException = FALSE;
+    _SEH2_TRY
+    {
+        pSheRemoveQuotesA(NULL);
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        bGotException = TRUE;
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/She.cpp:30]: Here is the `NULL` test.

File: dll/win32/shell32/stubs.cpp (hunk 1 above)
Comment (JoachimHenze) [reply dll/win32/shell32/stubs.cpp:460]: I don't think we should follow Microsofts behavior in crashing to such an extent. I do really vote for checking for NULL here. I don't think this would break any compatibility. I don't think there is any application out there that requires to crash "to behave correctly".

File: modules/rostests/apitests/shell32/She.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
+/*
+ * PROJECT:     ReactOS api tests
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Test for She* functions
```
Comment (HBelusca) [inline modules/rostests/apitests/shell32/She.cpp:4]: Is it going to serve for other ones (like SheSetCurDrive etc.), or only for the RemoveQuote functions?

File: dll/win32/shell32/stubs.cpp [hunk 5]
```diff
@@ -451,26 +451,38 @@ SheSetCurDrive(INT iIndex)
...
+    {
+        for (pch = psz + 1; *pch && *pch != '"'; ++pch)
+            *(pch - 1) = *pch;
```
Comment (HBelusca) [inline dll/win32/shell32/stubs.cpp:480]:
```suggestion
        {
            *(pch - 1) = *pch;
        }
```

File: modules/rostests/apitests/shell32/She.cpp (hunk 4 above)
Comment (katahiromz) [reply modules/rostests/apitests/shell32/She.cpp:4]: It's to serve for other ones.

File: dll/win32/shell32/stubs.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/stubs.cpp:460]: It should be "correctly crashing". That is the expected behavior.

Why does every PC crash on system failure? To stop the wrong process. To find the fatal mistake(s) on the system.

File: dll/win32/shell32/stubs.cpp [hunk 6]
```diff
@@ -451,26 +451,40 @@ SheSetCurDrive(INT iIndex)
...
+SheRemoveQuotesW(LPWSTR psz)
 {
-    FIXME("SheRemoveQuotesW() stub\n");
-    return NULL;
+    PWCHAR pch;
+
+    if (*psz == L'"')
+    {
+        for (pch = psz + 1; *pch && *pch != L'"'; ++pch)
+            *(pch - 1) = *pch;
```
Comment (HBelusca) [inline dll/win32/shell32/stubs.cpp:463]: for-loop braces here too (I didn't want to repeat my previous comment)

**Review by HBelusca — APPROVED** [review]
lgtm otherwise

File: dll/win32/shell32/stubs.cpp (hunk 6 above)
Comment (katahiromz) [reply dll/win32/shell32/stubs.cpp:464]: Fixed. Sorry.
